In [1]:
# Import the libraries 

import numpy as np 
import pandas as pd 
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib import style
style.use('ggplot')

In [2]:
# Importing the dataset 

test = pd.read_csv('test.csv')
train = pd.read_csv('train.csv')
sample = pd.read_csv('sample.csv')

In [3]:
# Shape of dataset 

test.shape, train.shape

((21805, 13), (50882, 14))

In [4]:
# Lets have a first look of train dataset 

train.head()

ID City_Code  Region_Code Accomodation_Type Reco_Insurance_Type  Upper_Age  \
0   1        C3         3213            Rented          Individual         36   
1   2        C5         1117             Owned               Joint         75   
2   3        C5         3732             Owned          Individual         32   
3   4       C24         4378             Owned               Joint         52   
4   5        C8         2190            Rented          Individual         44   

   Lower_Age Is_Spouse Health Indicator Holding_Policy_Duration  \
0         36        No               X1                     14+   
1         22        No               X2                     NaN   
2         32        No              NaN                     1.0   
3         48        No               X1                     14+   
4         44        No               X2                     3.0   

   Holding_Policy_Type  Reco_Policy_Cat  Reco_Policy_Premium  Response  
0                  3.0               22              11628.0         0  
1                  NaN               22              30510.0         0  
2                  1.0               19               7450.0         1  
3                  3.0               19              17780.0         0  
4                  1.0               16              10404.0         0

In [5]:
# Data Report function for Null and Unique Values 

def data_report(data):
    Null = pd.Series(data.isnull().sum())
    Unique_Count = pd.Series(data.describe(include='all',datetime_is_numeric=True).loc['unique', :])
    Data_type = pd.Series(data.dtypes)
    info_abt_data = pd.DataFrame(({"Null":Null, "Unique Count": Unique_Count, "Data type": Data_type}))
    return info_abt_data

In [6]:
# Print Data report fn 

data_report(train)

Null Unique Count Data type
ID                           0          NaN     int64
City_Code                    0           36    object
Region_Code                  0          NaN     int64
Accomodation_Type            0            2    object
Reco_Insurance_Type          0            2    object
Upper_Age                    0          NaN     int64
Lower_Age                    0          NaN     int64
Is_Spouse                    0            2    object
Health Indicator         11691            9    object
Holding_Policy_Duration  20251           15    object
Holding_Policy_Type      20251          NaN   float64
Reco_Policy_Cat              0          NaN     int64
Reco_Policy_Premium          0          NaN   float64
Response                     0          NaN     int64

In [7]:
data_report(test)

Null Unique Count Data type
ID                          0          NaN     int64
City_Code                   0           36    object
Region_Code                 0          NaN     int64
Accomodation_Type           0            2    object
Reco_Insurance_Type         0            2    object
Upper_Age                   0          NaN     int64
Lower_Age                   0          NaN     int64
Is_Spouse                   0            2    object
Health Indicator         5027            9    object
Holding_Policy_Duration  8603           15    object
Holding_Policy_Type      8603          NaN   float64
Reco_Policy_Cat             0          NaN     int64
Reco_Policy_Premium         0          NaN   float64

In [8]:
#ratio of null values
train.isnull().sum()/train.shape[0] *100

ID                          0.000000
City_Code                   0.000000
Region_Code                 0.000000
Accomodation_Type           0.000000
Reco_Insurance_Type         0.000000
Upper_Age                   0.000000
Lower_Age                   0.000000
Is_Spouse                   0.000000
Health Indicator           22.976691
Holding_Policy_Duration    39.799929
Holding_Policy_Type        39.799929
Reco_Policy_Cat             0.000000
Reco_Policy_Premium         0.000000
Response                    0.000000
dtype: float64

#### Ratio of null value is 22 and 39 %

In [9]:
#ratio of null values
test.isnull().sum()/test.shape[0] *100

ID                          0.000000
City_Code                   0.000000
Region_Code                 0.000000
Accomodation_Type           0.000000
Reco_Insurance_Type         0.000000
Upper_Age                   0.000000
Lower_Age                   0.000000
Is_Spouse                   0.000000
Health Indicator           23.054345
Holding_Policy_Duration    39.454254
Holding_Policy_Type        39.454254
Reco_Policy_Cat             0.000000
Reco_Policy_Premium         0.000000
dtype: float64

#### Null Values founded 
- Health Indicator	11691
- Holding_Policy_Duration	20251
- Holding_Policy_Type	20251

In [10]:
#categorical features
categorical = train.select_dtypes(include =[np.object])
print("Categorical Features in Train Set:",categorical.shape[1])

#numerical features
numerical= train.select_dtypes(include =[np.float64,np.int64])
print("Numerical Features in Train Set:",numerical.shape[1])

Categorical Features in Train Set: 6
Numerical Features in Train Set: 8


In [11]:
# Similar scenerio with test data also
# Lets move ahead to deal with train data's null values

-----------
----------
-----------

In [12]:
# renamed column name to deal with spacing issue further 

train = train.rename(columns= {"Health Indicator":'HealthIndicator'})
test = test.rename(columns= {"Health Indicator":'HealthIndicator'})

In [13]:
# before applying any imputation we have to deal with string values from of missing data columns

In [14]:
train['HealthIndicator']= train['HealthIndicator'].replace({'X1':9, 'X2':8, 'X3':7, 'X4':6, 'X5': 5, 'X6':4, 'X7':3 , 'X8':2, 'X9': 1 })
test['HealthIndicator']= test['HealthIndicator'].replace({'X1':9, 'X2':8, 'X3':7, 'X4':6, 'X5': 5, 'X6':4, 'X7':3 , 'X8':2, 'X9': 1 })

In [15]:
train['Holding_Policy_Duration']= train['Holding_Policy_Duration'].replace({'14+':15.0})
test['Holding_Policy_Duration']= test['Holding_Policy_Duration'].replace({'14+':15.0})

In [16]:
train.Holding_Policy_Duration.value_counts(normalize=True)

1.0     0.146877
15.0    0.141523
2.0     0.139075
3.0     0.117071
4.0     0.090464
5.0     0.077111
6.0     0.061833
7.0     0.053704
8.0     0.042963
9.0     0.036368
10.0    0.026542
11.0    0.017825
12.0    0.016748
13.0    0.016682
14.0    0.015213
Name: Holding_Policy_Duration, dtype: float64

In [1]:
train.Upper_Age.value_counts(normalize=True)

NameError: name 'train' is not defined

In [17]:
train.head(2)

ID City_Code  Region_Code Accomodation_Type Reco_Insurance_Type  Upper_Age  \
0   1        C3         3213            Rented          Individual         36   
1   2        C5         1117             Owned               Joint         75   

   Lower_Age Is_Spouse  HealthIndicator Holding_Policy_Duration  \
0         36        No              9.0                    15.0   
1         22        No              8.0                     NaN   

   Holding_Policy_Type  Reco_Policy_Cat  Reco_Policy_Premium  Response  
0                  3.0               22              11628.0         0  
1                  NaN               22              30510.0         0

In [18]:
# Looks like we are ready to apply imputer 

In [19]:
# I'm applying KNN imputer for dealing with null values

from sklearn.impute import KNNImputer
imputer = KNNImputer(n_neighbors=2)

train["HealthIndicator"] = pd.DataFrame(imputer.fit_transform(train[["HealthIndicator"]]))
train["Holding_Policy_Duration"] = pd.DataFrame(imputer.fit_transform(train[["Holding_Policy_Duration"]]))
train["Holding_Policy_Type"] = pd.DataFrame(imputer.fit_transform(train[["Holding_Policy_Type"]]))

test["HealthIndicator"] = pd.DataFrame(imputer.fit_transform(test[["HealthIndicator"]]))
test["Holding_Policy_Duration"] = pd.DataFrame(imputer.fit_transform(test[["Holding_Policy_Duration"]]))
test["Holding_Policy_Type"] = pd.DataFrame(imputer.fit_transform(test[["Holding_Policy_Type"]]))

In [20]:
# now lets check the Data report 

data_report(train)

Null Unique Count Data type
ID                          0          NaN     int64
City_Code                   0           36    object
Region_Code                 0          NaN     int64
Accomodation_Type           0            2    object
Reco_Insurance_Type         0            2    object
Upper_Age                   0          NaN     int64
Lower_Age                   0          NaN     int64
Is_Spouse                   0            2    object
HealthIndicator             0          NaN   float64
Holding_Policy_Duration     0          NaN   float64
Holding_Policy_Type         0          NaN   float64
Reco_Policy_Cat             0          NaN     int64
Reco_Policy_Premium         0          NaN   float64
Response                    0          NaN     int64

In [21]:
test_data_id = test['ID']

test_data_id.shape

(21805,)

#### Lets move further to deal with categorical values

#### Column having categorical data 
- City_code 
- Accomodation_Type
- Reco_Insurance_Type
- Is_Spouse

In [22]:
train['City_Code'].unique()

array(['C3', 'C5', 'C24', 'C8', 'C9', 'C1', 'C15', 'C28', 'C27', 'C7',
       'C20', 'C25', 'C4', 'C2', 'C34', 'C10', 'C17', 'C18', 'C16', 'C29',
       'C33', 'C26', 'C19', 'C6', 'C12', 'C13', 'C11', 'C14', 'C22',
       'C23', 'C21', 'C36', 'C32', 'C30', 'C35', 'C31'], dtype=object)

In [23]:
train['Reco_Insurance_Type'].unique()

array(['Individual', 'Joint'], dtype=object)

In [24]:
train['Accomodation_Type'].unique()

array(['Rented', 'Owned'], dtype=object)

In [25]:
train['Is_Spouse'].unique()

array(['No', 'Yes'], dtype=object)

In [26]:
train['Is_Spouse']= train['Is_Spouse'].replace({'No':0, 'Yes':1})
test['Is_Spouse']= test['Is_Spouse'].replace({'No':0, 'Yes':1})

In [27]:
train.head(3)

ID City_Code  Region_Code Accomodation_Type Reco_Insurance_Type  Upper_Age  \
0   1        C3         3213            Rented          Individual         36   
1   2        C5         1117             Owned               Joint         75   
2   3        C5         3732             Owned          Individual         32   

   Lower_Age  Is_Spouse  HealthIndicator  Holding_Policy_Duration  \
0         36          0         9.000000                15.000000   
1         22          0         8.000000                 6.156769   
2         32          0         7.555306                 1.000000   

   Holding_Policy_Type  Reco_Policy_Cat  Reco_Policy_Premium  Response  
0             3.000000               22              11628.0         0  
1             2.439228               22              30510.0         0  
2             1.000000               19               7450.0         1

In [28]:
train.Reco_Insurance_Type.value_counts(normalize=True)

Individual    0.796667
Joint         0.203333
Name: Reco_Insurance_Type, dtype: float64

In [29]:
train.Accomodation_Type.value_counts(normalize=True)

Owned     0.54933
Rented    0.45067
Name: Accomodation_Type, dtype: float64

In [30]:
train.City_Code.value_counts(normalize=True)

C1     0.176349
C2     0.152254
C3     0.096085
C4     0.072147
C9     0.042942
C6     0.038324
C7     0.037538
C8     0.035494
C10    0.031661
C5     0.026807
C15    0.023309
C17    0.022778
C11    0.022542
C16    0.022307
C13    0.019752
C20    0.018199
C19    0.017668
C12    0.017059
C18    0.015664
C14    0.014661
C21    0.013345
C23    0.011536
C24    0.010868
C22    0.010141
C26    0.009807
C29    0.007606
C25    0.007193
C27    0.005798
C33    0.005621
C28    0.005601
C32    0.003145
C34    0.002555
C30    0.001140
C35    0.001101
C36    0.000708
C31    0.000295
Name: City_Code, dtype: float64

In [31]:
train = pd.get_dummies(train, columns=["Reco_Insurance_Type", "Accomodation_Type"])
test = pd.get_dummies(test, columns=["Reco_Insurance_Type", "Accomodation_Type"])

In [32]:
train.head(3)

ID City_Code  Region_Code  Upper_Age  Lower_Age  Is_Spouse  \
0   1        C3         3213         36         36          0   
1   2        C5         1117         75         22          0   
2   3        C5         3732         32         32          0   

   HealthIndicator  Holding_Policy_Duration  Holding_Policy_Type  \
0         9.000000                15.000000             3.000000   
1         8.000000                 6.156769             2.439228   
2         7.555306                 1.000000             1.000000   

   Reco_Policy_Cat  Reco_Policy_Premium  Response  \
0               22              11628.0         0   
1               22              30510.0         0   
2               19               7450.0         1   

   Reco_Insurance_Type_Individual  Reco_Insurance_Type_Joint  \
0                               1                          0   
1                               0                          1   
2                               1                          0   

   Accomodation_Type_Owned  Accomodation_Type_Rented  
0                        0                         1  
1                        1                         0  
2                        1                         0

In [ ]:
# generate binary values using get_dummies
# train = pd.get_dummies(train, columns=["City_Code"], prefix=["Type_is"] )

In [35]:
test.shape, train.shape

((21805, 15), (50882, 16))

In [38]:
test.to_csv('cleantest.csv', index=False)

In [39]:
train.to_csv('cleantrain.csv', index=False)